### ISSUE TITLE
Significantly slower performance when using LLVM for json2msgpack compared to GCC

### ISSUE THREAD

**Issue by lemire** [simdjson/simdjson#1910]
LLVM 14:

```
$ cmake --build buildclang --target bench_ondemand && ./buildclang/benchmark/bench_ondemand --benchmark_filter="json2msgpack<simdjson_ondemand>"
Running ./buildclang/benchmark/bench_ondemand
Run on (128 X 851.01 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x64)
  L1 Instruction 32 KiB (x64)
  L2 Unified 1280 KiB (x64)
  L3 Unified 55296 KiB (x2)
Load Average: 0.22, 0.10, 0.09
simdjson::dom implementation:      icelake
simdjson::ondemand implementation (stage 1): icelake
simdjson::ondemand implementation (stage 2): fallback
------------------------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------------------
json2msgpack<simdjson_ondemand>/manual_time     483451 ns       606029 ns         1456 best_branch_miss=1.571k best_bytes_per_sec=1.31731G best_cache_miss=0 best_cache_ref=23.97k best_cycles=1.63043M best_cycles_per_byte=2.58177 best_docs_per_sec=2.08596k best_frequency=3.401G best_instructions=3.08459M best_instructions_per_byte=4.88443 best_instructions_per_cycle=1.89189 best_items_per_sec=2.08596k branch_miss=1.63967k bytes=631.515k bytes_per_second=1.21655G/s cache_miss=0.285027 cache_ref=24.0078k cycles=1.64337M cycles_per_byte=2.60226 docs_per_sec=2.06846k/s frequency=3.39924G/s instructions=3.08459M instructions_per_byte=4.88443 instructions_per_cycle=1.87699 items=1 items_per_second=2.06846k/s [BEST: throughput=  1.32 GB/s doc_throughput=  2085 docs/s instructions=     3084589 cycles=     1630428 branch_miss=    1571 cache_miss=       0 cache_ref=     23970 items=         1 avg_time=    483451 ns]
```

GCC 12:

```
$ cmake --build buildgcc12 --target bench_ondemand && ./buildgcc12/benchmark/bench_ondemand --benchmark_filter="json2msgpack<simdjson_ondemand>"
2022-10-12T22:04:33+00:00
Running ./buildgcc12/benchmark/bench_ondemand
Run on (128 X 1004.66 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x64)
  L1 Instruction 32 KiB (x64)
  L2 Unified 1280 KiB (x64)
  L3 Unified 55296 KiB (x2)
Load Average: 0.39, 0.16, 0.11
simdjson::dom implementation:      icelake
simdjson::ondemand implementation (stage 1): icelake
simdjson::ondemand implementation (stage 2): fallback
------------------------------------------------------------------------------------------------------
Benchmark                                            Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------------------
json2msgpack<simdjson_ondemand>/manual_time     232907 ns       356095 ns         2995 best_branch_miss=957 best_bytes_per_sec=2.73422G best_cache_miss=0 best_cache_ref=24.561k best_cycles=785.774k best_cycles_per_byte=1.24427 best_docs_per_sec=4.32962k best_frequency=3.40211G best_instructions=2.80056M best_instructions_per_byte=4.43467 best_instructions_per_cycle=3.56408 best_items_per_sec=4.32962k branch_miss=978.318 bytes=631.515k bytes_per_second=2.52523G/s cache_miss=1.59967 cache_ref=24.5545k cycles=792.002k cycles_per_byte=1.25413 docs_per_sec=4.29356k/s frequency=3.40051G/s instructions=2.80056M instructions_per_byte=4.43467 instructions_per_cycle=3.53605 items=1 items_per_second=4.29356k/s [BEST: throughput=  2.73 GB/s doc_throughput=  4329 docs/s instructions=     2800560 cycles=      785774 branch_miss=     957 cache_miss=       0 cache_ref=     24561 items=         1 avg_time=    232906 ns]
```

It is a 2x difference on this one benchmark.

Other benchmarks do not show such large differences. It is worth investigating !!!

**Comment by jkeiser** [comment]
The thing that distinguishes json2msgpack from other On Demand cases is recursion. Anything that generically walks JSON has a radically different shape than typical cases where the JSON structure is known ahead of time. An optimizer that can't see through recursive function calls may end up storing the iterators in memory instead of in registers, which makes a ridiculously massive difference. Just a shot from the hip.

**Comment by lemire** [comment]
You may be right, but GCC is not massively smarter than LLVM clang. In any case, it is worth investigating !

**Comment by TysonAndre** [comment]
What are the optimization flags being used?

In gcc, I think that `-O3` would include a link-time optimization pass by default. In clang (only have clang 10 installed, haven't checked if this changed in clang 12), "-fno-lto" is the default.

(the benchmarks would make a lot of calls to methods implemented in simdjson.cpp with the ondemand parser)

Separately, I think that flto could be better documented for compilers that don't enable it by default - https://github.com/simdjson/simdjson/search?q=%22flto%22 it's only mentioned in the issue tracker

- Does it make sense to automatically enable this for clang when building this as a standalone shared library or running benchmarks


----

Brainstorming: I guess default inlining depth limits may also be relevant, but haven't checked if it's inlining the recursive call

**Comment by jkeiser** [comment]
One note: I think the only call to simdjson.cpp is the one to stage 1. The entire frontend (including string and number parsing) is inline in .h files.

**Comment by lemire** [comment]
@jkeiser You are correct except for one function call which unescape strings. It is something new I added. However, I am quite confident that it is not (on its own) an explanation because even if you skip this step (do not process strings), the clang benchmark is still slower than the gcc benchmark (even if the gcc binary does process the strings).

Conceptually, we can include dispatched functions as needed, and I found that it was beneficial to dispatch the string processing function.
